### PR TITLE
let the rules review endpoint respect scopes

### PR DESCRIPTION
**Pull request by deads2k**
Adds a `Scopes` field for a self subject rules review.

@openshift/api-review

### CODE REVIEW

File: pkg/authorization/api/types.go [hunk 1]
```diff
@@ -138,10 +138,21 @@ type PolicyBinding struct {
...
 	// Status is completed by the server to tell which permissions you have
 	Status SubjectRulesReviewStatus
 }
 
+// SubjectRulesReviewSpec adds information about how to conduct the check
+type SubjectRulesReviewSpec struct {
+	// Scopes to use for the evaluation.  Empty means "use the unscoped (full) permissions of the user/groups".
+	// Nil for a self-SubjectRulesReview, means "use the scopes on this request".
+	// Nil for a regular SubjectRulesReview, means the same as empty.
+	Scopes []string
```
Comment (deads2k) [inline pkg/authorization/api/types.go]: Same rules as for SAR.

File: pkg/authorization/authorizer/scope/converter.go [hunk 2]
```diff
@@ -126,6 +126,7 @@ func (userEvaluator) ResolveRules(scope, namespace string, clusterPolicyGetter r
 	case UserIndicator + UserAccessCheck:
 		return []authorizationapi.PolicyRule{
 			{Verbs: sets.NewString("create"), APIGroups: []string{authorizationapi.GroupName}, Resources: sets.NewString("subjectaccessreviews", "localsubjectaccessreviews"), AttributeRestrictions: &authorizationapi.IsPersonalSubjectAccessReview{}},
+			authorizationapi.NewRule("create").Groups(authorizationapi.GroupName).Resources("selfsubjectrulesreviews").RuleOrDie(),
```
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: `check-access` should let you do the rules review.

File: pkg/authorization/authorizer/scope/converter.go [hunk 3]
```diff
@@ -214,8 +215,7 @@ func (e clusterRoleEvaluator) ResolveRules(scope, namespace string, clusterPolic
 		return []authorizationapi.PolicyRule{}, nil
 	}
 
-	ctx := kapi.WithNamespace(kapi.NewContext(), namespace)
```
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: No namespaces are allowed for cluster policy lookups.

**Comment by deads2k** [comment]
builds on https://github.com/openshift/origin/pull/9106

File: pkg/authorization/authorizer/scope/converter.go (hunk 3 above)
Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: how was this working before?

File: pkg/authorization/registry/selfsubjectrulesreview/storage.go [hunk 4]
```diff
@@ -22,12 +28,48 @@ func (r *REST) New() runtime.Object {
...
+	namespace := kapi.NamespaceValue(ctx)
+	if len(namespace) == 0 {
+		return nil, kapierrors.NewBadRequest(fmt.Sprintf("namespace is required on this type: %v", namespace))
+	}
+	user, exists := kapi.UserFrom(ctx)
+	if !exists {
+		return nil, fmt.Errorf("user missing from context")
+	}
+
 	policyRules, err := r.ruleResolver.GetEffectivePolicyRules(ctx)
```
Comment (liggitt) [inline pkg/authorization/registry/selfsubjectrulesreview/storage.go]: previously, would this let you check your cluster-level access as well?

File: pkg/authorization/api/v1beta3/types.go [hunk 5]
```diff
@@ -113,10 +113,21 @@ type NamedRoleBinding struct {
...
+	Spec SubjectRulesReviewSpec `json:"status`
+
 	// Status is completed by the server to tell which permissions you have
 	Status SubjectRulesReviewStatus `json:"status,omitempty"`
 }
 
+// SubjectRulesReviewSpec adds information about how to conduct the check
+type SubjectRulesReviewSpec struct {
+	// Scopes to use for the evaluation.  Empty means "use the unscoped (full) permissions of the user/groups".
+	// Nil for a self-SubjectRulesReview, means "use the scopes on this request".
```
Comment (smarterclayton) [inline pkg/authorization/api/v1beta3/types.go]: This doc probably belongs in each type, not in the spec itself.  I.e. don't define the behavior in the struct that is polymorphic

File: pkg/authorization/registry/selfsubjectrulesreview/storage.go [hunk 6]
```diff
@@ -22,12 +28,48 @@ func (r *REST) New() runtime.Object {
...
+	rules := []authorizationapi.PolicyRule{}
+	for _, rule := range policyRules {
+		rules = append(rules, rulevalidation.BreakdownRule(rule)...)
```
Comment (liggitt) [inline pkg/authorization/registry/selfsubjectrulesreview/storage.go]: move the breakdown into `filterRulesByScopes`? or did you want the collation `CompactRules` can do for us with simple rules?

Comment (liggitt) [inline pkg/authorization/registry/selfsubjectrulesreview/storage.go]: actually, doesn't the Covers check already break down the rules? just wondering how expensive this API call is going to be

Comment (deads2k) [inline pkg/authorization/registry/selfsubjectrulesreview/storage.go]: > move the breakdown into filterRulesByScopes? or did you want the collation CompactRules can do for us with simple rules?

I want the collation, so I need the broken down rules anyway.

Comment (deads2k) [inline pkg/authorization/registry/selfsubjectrulesreview/storage.go]: > actually, doesn't the Covers check already break down the rules? just wondering how expensive this API call is going to be

It breaks down the needle, but if the needle is already broken down, that's a reasonably fast no-op.  No net allocations compared to the alternative, maybe one extra pass through a loop?

File: pkg/authorization/registry/selfsubjectrulesreview/storage.go (hunk 4 above)
Comment (deads2k) [inline pkg/authorization/registry/selfsubjectrulesreview/storage.go]: > previously, would this let you check your cluster-level access as well?

No, it was a namespace scoped resource, so there was no way to call it like that.

File: pkg/authorization/authorizer/scope/converter.go (hunk 3 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: > how was this working before?

I think the getter against the cache ignored the field.  This is uncached until we get SharedInformers, so I don't have to plumb it through twice.

**Comment by liggitt** [comment]
reviewed, no other comments from me

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
yum re[test]

**Comment by deads2k** [comment]
conformance hang, re[test]

**Comment by deads2k** [comment]
conformance, yum re[merge]

**Comment by deads2k** [comment]
verify unrelated to me [merge]

**Pull request merged**